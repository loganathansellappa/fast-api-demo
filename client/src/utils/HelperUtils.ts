import { toast } from 'react-toastify';

export type ServerConfig = {
    baseUrl: string;
};
export const serverData = () => {
    return {
        baseUrl: process.env.DATA_URL,
    } as ServerConfig;
};

export enum MissionState {
    PreFlight = 'Pre-flight',
    Flight = 'In-Flight',
    PostFlight = 'Post-Flight'
}


export enum MutationType {
    PUT = 'PUT',
    POST = 'POST',
    DELETE = 'DELETE'
}

export const showSuccessToast = (message: string, id: string | number = 1) => {
    toast.dismiss();
    toast.success(message, { toastId: id, autoClose: 500, style: { background: '#5CDD00', color: "white" } });
};

export const showErrorToast = (message: string, id: string | number) => {
    toast.dismiss();
    toast.error(message, { toastId: id, autoClose: 500 });
};
