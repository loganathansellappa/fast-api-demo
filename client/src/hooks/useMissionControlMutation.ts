import { useAxiosClient } from '../context/AxiosClientContext.ts';
import { useMutation, useQueryClient } from 'react-query';
import { missionControlKeys } from '../services/queryKeyFactory.ts';
import { MissionState, MutationType } from '../utils/HelperUtils.ts';

export type MissionControlMutation = {
  missionState?: MissionState;
  mutationType: MutationType;
  url: string;
  description?: string;
  title?: string;
};

export type MissionControlInput = {
  mutationType: MutationType;
  id?: number;
  missionState?: MissionState;
  mission_state?: MissionState;
  description?: string;
  title?: string;
};

type RequestData = {
  url: string;
  method: MutationType;
  mission_state?: MissionState;
  data?: Record<string, unknown>;
};

export function useMissionControlUpdate() {
  const axiosClientService = useAxiosClient();
  const queryClient = useQueryClient();

  const mutation = useMutation({
    mutationFn: (requestData: MissionControlMutation) => {
      const { url, mutationType, missionState, ...data } = requestData;
      const requestObject: RequestData = {
          url: url,
          method: mutationType
      }
      if (mutationType !== MutationType.DELETE) {
        requestObject.data  = {
          ...data,
          mission_state: missionState
        }
      }
      return axiosClientService.axiosClient.request({...requestObject })
    },
    onSuccess: () => {
      queryClient.cancelQueries({ queryKey: missionControlKeys.all() });
      queryClient.invalidateQueries({ queryKey: missionControlKeys.all() });
    },
    onError: (error) => {
      console.error(error);
    }
  });

  const mutateMissionControl = ({ id, missionState = MissionState.PreFlight, mutationType, description, title }: MissionControlInput) => {
    const url = mutationType === MutationType.POST ? `flight_missions` : `flight_missions/${id}`;
    if (mutationType === MutationType.DELETE) {
      mutation.mutate({ mutationType, url });
    }  else {
      mutation.mutate({ missionState, mutationType, description, title, url });
    }

  };

  return {
    mutate: mutateMissionControl,
    isLoading: mutation.isLoading,
    isError: mutation.isError,
    error: mutation.error,
    isSuccess: mutation.isSuccess,
  };
}
