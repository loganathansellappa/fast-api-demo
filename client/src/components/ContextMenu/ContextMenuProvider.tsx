import React, { useCallback, useMemo, useState } from 'react';
import { ContextMenu } from "./ContextMenu";
import { MissionState, MutationType, showErrorToast, showSuccessToast } from '../../utils/HelperUtils.ts';
import { MissionControlInput, useMissionControlUpdate } from '../../hooks/useMissionControlMutation.ts';
export interface ContextMenuProviderProps {
  children: React.ReactNode;
  options: MissionState[],
}

export const ContextMenuProvider: React.FC<ContextMenuProviderProps> = ({
                                                                          children,
                                                                          options,
                                                                        }) => {
  const [contextMenu, setContextMenu] = useState<{
    show: boolean;
    top: number;
    left: number;
    item?: number;
  }>({
    show: false,
    top: 0,
    left: 0,
    item: 0,
  });
  const {  mutate, isError, isSuccess} = useMissionControlUpdate();


  const handleClick = useCallback((currentItem: number, e: React.MouseEvent<HTMLDivElement, MouseEvent>) => {
    const target = e.target as HTMLElement;
    const currentState = target.innerHTML as MissionState;
    const requestData: MissionControlInput = {
      id: currentItem as number,
      missionState: currentState,
      mutationType: MutationType.PUT,
    }
    mutate(requestData);
  },[mutate])


  if(isError) {
    showErrorToast('Mission Control state update failed', `missionControlStateUpdateFailed$`);
  }

  if(isSuccess) {
    showSuccessToast('Mission Control state updated', `missionControlStateUpdateSuccess`);
  }

  const menuOptions = useMemo(() => Object.values(options).map(state => state),[options]);
  const handleContextMenu = (e: React.MouseEvent<HTMLDivElement>) => {
    e.preventDefault();
    e.stopPropagation();
    const element = e.target as HTMLElement;
    setContextMenu({ show: true, top: e.clientY, left: e.clientX, item: element.dataset.itemId as unknown as number});
  };

  const handleContextMenuClose = () => {
    setContextMenu({ show: false, top: 0, left: 0, item: 0});
  };

  return (
    <div onContextMenu={handleContextMenu} onClick={handleContextMenuClose}>
      {children}
      <ContextMenu
        showMenu={contextMenu.show}
        position={{ top: contextMenu.top, left: contextMenu.left }}
        currentItem={contextMenu.item as number}
        options={menuOptions}
        handleClick={handleClick}
      />
    </div>
  );
};
