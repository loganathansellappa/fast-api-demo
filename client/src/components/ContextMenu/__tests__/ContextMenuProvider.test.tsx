import React from 'react';
import { render } from '@testing-library/react';
import { MissionState} from '../../../utils/HelperUtils.ts';
import { ContextMenuProvider, ContextMenuProviderProps } from '../ContextMenuProvider.tsx';
import * as Hooks from '../../../hooks/useMissionControlMutation.ts';


describe('ContextMenuProvider', () => {
  let element: { container: HTMLElement };
  const mutateMissionControl = jest.fn();

  afterEach(() => {
    jest.clearAllMocks();
  });

  beforeEach(() => {
    jest
      .spyOn(Hooks, "useMissionControlUpdate")
      .mockReturnValue(
        {
          mutate: mutateMissionControl,
          isLoading: false,
          isError: false,
          error: false,
          isSuccess: false,
        }
      );
    const props: ContextMenuProviderProps = {
      children: React.createElement('div', { className: 'dummy-child', 'data-testid': 'child' }),
      options: Object.values(MissionState).map(state => state)
    };
    element = render(<ContextMenuProvider {...props} />);
  });

  test("renders ContextMenuProvider component", () => {
    expect(element.container.firstChild).toMatchSnapshot();
  });

  test("renders children ", () => {
    expect(
      element.container.getElementsByClassName("dummy-child")!.length,
    ).toBe(1);
  });
});
