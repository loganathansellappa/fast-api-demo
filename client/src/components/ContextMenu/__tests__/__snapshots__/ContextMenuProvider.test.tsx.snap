// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContextMenuProvider renders ContextMenuProvider component 1`] = `
<div>
  <div
    class="dummy-child"
    data-testid="child"
  />
</div>
`;
